### ISSUE TITLE
Distributed Nodes

### ISSUE THREAD

**Issue by slawlor** [slawlor/ractor#16]
Following from Erlang, we want to support distributed nodes
https://www.erlang.org/doc/reference_manual/distributed.html

We need

- [ ] Tcp connections and transmission
- [ ] Encrypted socket communications
- [ ] Dist specification
- [ ] Serializable messages which can go over the network boundary in a distributed environment
- [ ] Each "node" needs to know of all nodes it's dist connected to (this includes updating the local PG groups and named registry with remote nodes)
- [ ] Each "remote" node should be a local actor that talks to the remote channel, allowing for local message processing without knowing anything about the actor being remote
- [ ] Probably a bunch more stuff...

We have:
- [x] We added support for remote referenced actor ids in #14 [see this commit](https://github.com/slawlor/ractor/pull/14/commits/796df54f540530172e857a35373b5caec529057f#diff-8179e798a78fff77fb35233cebe7cce8852fcb9a4226546d295d92a1cfd5d352)

**Comment by slawlor** [comment]
Related: #39 #36

**Comment by 0x53A** [comment]
Hey I hope it's alright if I just barge in!

From this issue it looks like ractor_cluster is a mesh cluster where every node can communicate with every other node directly, is that correct?

I think a star topology would also be extremely interesting for other scenarios, like adding microcontrollers or web browsers.

In both of these cases, there would be a 1:1 connection between a dedicated server and client. The client can transparently call actors on the server through proxy objects, and vice versa. If there's a cluster with multiple nodes on the server side, the server would handle routing the message to the correct recipient.

WASM <-> Server communication would require a websocket transport (or http/2 magic).

Together with a cross platform UI library like egui that would be my dream for having truly portable applications.

**Comment by slawlor** [comment]
@0x53A 

> From this issue it looks like ractor_cluster is a mesh cluster where every node can communicate with every other node directly, is that correct?

That's correct, it's true-to-form for Erlang clustering paradigms 

> I think a star topology would also be extremely interesting for other scenarios, like adding microcontrollers or web browsers.

For sure! But networking configurations are particular tricky especially with reconnects/etc and I just honestly haven't had the time to invest more here.

Feel free to contributed or put up changes you would like to investigate. We're very open to new contributors

**Comment by 0x53A** [comment]
As a small spoiler, I've been working on exactly that the last few days. The work-in-progress repo is here, though it's not yet functional: https://github.com/0x53A/ractor_test

**Comment by chenyukang** [comment]
@slawlor thanks for this great project and reference!
We have a project which heavily using ractor as framework. Basically, this is a distributed system with high demands for correctness and safety (you can think of it as a distributed database). During the actual development process, I spent a lot of time debugging issues caused by distributed systems—some bugs couldn't be reliably reproduced, or they only appeared on certain machines. 

To ensure product quality, we want to introduce fuzz testing. I looked at the code of ractor, and it seems there’s no support for this. Do you have any thought or recommended tool on fuzz testing? Perhaps we can add some related functionality to ractor. 

The simplest approach I can think of is to add simulated delays on critical code paths to mimic real-world scenarios.

**Comment by slawlor** [comment]
> To ensure product quality, we want to introduce fuzz testing. I looked at the code of ractor, and it seems there’s no support for this. Do you have any thought or recommended tool on fuzz testing? Perhaps we can add some related functionality to ractor.

Yeah this wasn't on my roadmap, but I imagine you could try and put a hook into the serialization calls so when messages serialize you could setup fuzzing on the results? But it wouldn't be a trivial hook, and I haven't really considered how it'd go.

I'm very open to any contributions if you can figure out a good way to hook into it, you have my full support!
